### PR TITLE
Support W3C incoming headers in opt-in mode

### PR DESCRIPTION
**Pull request by lmolkova**
Implements [w3c distributed tracing standard](https://github.com/w3c/distributed-tracing).

This change enables opt-in W3C support via settings.
To enable it, set `ApplicationInsightsServiceOptions.RequestCollectionOptions.EnableW3CDistributedTracing` to true and provide options to the `AddApplicationInsights telemetry`

This change ensures W3C ids are set on the telemetry (operationId, parentId and Id). 'legacy' ids are set in custom dimensions whenever they are different from the W3C ones.

This change requires UI queries change to support querying for custom dimensions for the cases when operation id or parent id do not match on the legacy-W3C boundary.

~~This PR depends on changes in DependencyCollector that will be shipped in 2.8.0-beta1 (https://github.com/Microsoft/ApplicationInsights-dotnet-server/pull/945) and cannot be merged until that. Sharing it to get the early feedback.~~

@cijothomas @reyang @SergeyKanzhelev please review

### CODE REVIEW

**Comment by lmolkova** [comment]
ready for review @cijothomas ;)

File: test/Microsoft.ApplicationInsights.AspNetCore.Tests/RequestTrackingMiddlewareTest.cs [hunk 1]
```diff
@@ -581,6 +588,237 @@ public void DoesntAddSourceIfRequestHeadersDontHaveSource()
...
+            Assert.Equal($"|4bf92f3577b34da6a3ce929d0e0e4736.{activityInitializedByW3CHeader.GetSpanId()}.", requestTelemetry.Id);
+            Assert.Equal("4bf92f3577b34da6a3ce929d0e0e4736", requestTelemetry.Context.Operation.Id);
+            Assert.Equal("|4bf92f3577b34da6a3ce929d0e0e4736.00f067aa0ba902b7.", requestTelemetry.Context.Operation.ParentId);
+
+            Assert.True(context.Response.Headers.TryGetValue(RequestResponseHeaders.RequestContextHeader, out var appId));
+            Assert.Equal($"appId={CommonMocks.TestApplicationId}", appId);
+        }
+
+        [Fact]
+        public void OnHttpRequestInStartWithW3CHeadersAndRequestIdIsTrackedCorrectly()
```
Comment (cijothomas) [inline test/Microsoft.ApplicationInsights.AspNetCore.Tests/RequestTrackingMiddlewareTest.cs]: rename unit test to indicate OnBeginRequest bein tested

File: src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs [hunk 2]
```diff
@@ -138,36 +173,54 @@ public void OnBeginRequest(HttpContext httpContext, long timestamp)
...
+                string sourceAppId = null;
+
                 IHeaderDictionary requestHeaders = httpContext.Request.Headers;
-                if (requestHeaders.TryGetValue(RequestResponseHeaders.RequestIdHeader, out requestId))
+
+                string originalParentId = null;
+                // W3C
+                if (this.enableW3CHeaders)
+                {
+                    SetW3CContext(httpContext.Request.Headers, activity, out sourceAppId);
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs]: I like this idea, I also wonder if it is secure?

[the remark this answers is not in the document]
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs]: AS discussed, lets not do it now. We can ask Cx to add TI to dump headers, if a need arises.

**Pull request merged**